### PR TITLE
Mini projects list update

### LINKED ISSUE #4111: Add starter projects from wiki to mini-projects list
As discussed in #3930: it would seem like a good idea to add several projects from https://web.archive.org/web/20150530010030/http://www.cprover.org/wiki/doku.php?id=starter_projects to the in-tree projects list, possibly introducing several levels (from actual "mini" projects to larger ones).

### PR DESCRIPTION
**Pull request by NlightNFotis**
This fixes #4111. Port of a list of starter projects from the old CPROVER wiki
into the in-repo file tracking them.

- [ ] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [x] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

**Comment by NlightNFotis** [comment]
Note to reviewers: While porting this list, I found a reference to a lot of names, or WIP notes that I couldn't judge the validity of. These are marked as `(?)` in the file, and I would appreciate if someone could tell me if they are valid or not any more.

Also while porting the list, I removed references to a couple of projects like the DALVIK one, judging that it's no longer relevant. If someone thinks otherwise, I'm happy to add it back.

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: df9625add6bc9f0f98705724268b3557e747312f).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/100110943

**Comment by NlightNFotis** [comment]
@tautschnig Could I please get a review on this and get it merged if it looks good?

File: MINI-PROJECTS.md [hunk 1]
```diff
@@ -104,6 +104,314 @@ share a bound.
...
+would add a new instrumentation option for `cbmc` and `goto-instrument` that checks
+each pointer before dereference and makes sure it's correctly aligned.
+
+### Command line options to SAT solvers
+
+There are options for which SMT solver to run and there are ways of building with
+multiple SAT solvers but no command line options to pick a SAT solver. It would be
+nice if there were.
+
+César is working on this one. (?)
```
Comment (tautschnig) [inline MINI-PROJECTS.md]: I think we also have a stalled pull request (#1501 ?) on that one.

File: MINI-PROJECTS.md [hunk 2]
```diff
@@ -104,6 +104,314 @@ share a bound.
...
+
+### Combine Uninitialize Variables Analysis with Alias Analysis
```
Comment (tautschnig) [inline MINI-PROJECTS.md]: Uninitialized

File: MINI-PROJECTS.md [hunk 3]
```diff
@@ -104,6 +104,314 @@ share a bound.
...
+
+`goto-instrument` features an instrumentation to track cases of use-before-define, i.e.
+reading of uninitialised variables. At present, the analysis does not take into account
+aliasing information. As such, the following code will yield a spurious counterexample:
```
Comment (tautschnig) [inline MINI-PROJECTS.md]: I actually think that instrumentation could do with some initial work just to make sure it works at all :-)

File: MINI-PROJECTS.md [hunk 4]
```diff
@@ -104,6 +104,314 @@ share a bound.
...
+}
+```
+
+### Constant Folding
+
+With the recently added reaching definitions analysis, it is straightforward to implement
+a program transformation that performs constant folding. ~To ensure soundness in concurrent
+settings, it should be combined with the existing is_threaded analysis.~ (A concurrency-aware
+version of reaching definitions now exists and is awaiting a merge.) Along the way, assertions
+that become trivially true should be eliminated.
```
Comment (tautschnig) [inline MINI-PROJECTS.md]: I believe constant folding is actually implemented to the degree described here.

File: MINI-PROJECTS.md [hunk 5]
```diff
@@ -104,6 +104,314 @@ share a bound.
...
+
+Re-implement the algorithm described in [this paper](http://www.kroening.com/papers/STTT-error-explanation-2005.pdf)
+
+### LTL
+
+Build an instrumenter in `goto-instrument` for checking LTL (or even CTL) properties on C code.
+E.g look at [this paper](https://www.umsec.umn.edu/publications/Partial-Translation-Verification-Untrusted-Code)
+(And please get in touch with Michael Tautschnig, who has done very similar work in FShell).
+
+Rajdeep is now looking into this one. (?)
```
Comment (tautschnig) [inline MINI-PROJECTS.md]: No idea whether that's still a live project.

File: MINI-PROJECTS.md [hunk 6]
```diff
@@ -104,6 +104,314 @@ share a bound.
...
+
+### Test SMT Back Ends
+
+CBMC allows to choose among a selection of SMT solvers to use as solving back end. As these
+tools evolve, CBMC's code for running them may need a fresh look from time to time. After
+a first round of manual experiments, these tests should become part of the routine build
+process on `dkr-build`. (Michael Tautschnig knows how to do this.)
```
Comment (tautschnig) [inline MINI-PROJECTS.md]: Today we've got proper CI, no need to resort do `dkr-build`.

File: MINI-PROJECTS.md [hunk 7]
```diff
@@ -104,6 +104,314 @@ share a bound.
...
+
+### Simple Test Harness Generation
+
+CBMC currently outputs a trace of execution when it find a counter-example. Instead it
+should be possible to produce a C file that when compiled and run will cause the program
+to execute this trace. This will allow a faster verify/debug cycle. (To a certain
+extent this has been accomplished in FShell. Speak to Michael.)
+
+Pascal is now looking at this one. (?)
```
Comment (tautschnig) [inline MINI-PROJECTS.md]: I guess there is an actual product for Java nowadays :-)

File: MINI-PROJECTS.md [hunk 8]
```diff
@@ -104,6 +104,314 @@ share a bound.
...
+
+### Improved Interpreter
+
+CProver has an interpreter for goto programs but it could be improved, for example adding
+support for dynamic analysis, calls to external functions and possibly even mixed concrete
+and symbolic execution. It would also be good to be able to interpret traces found by BMC
+to catch mismatches / bugs between the logical model and the bit blasting models. This
+would also be useful for more advanced CEGAR like techniques (also see below).
```
Comment (tautschnig) [inline MINI-PROJECTS.md:169]: I have no idea what the actual state of the interpreter is today.

File: MINI-PROJECTS.md [hunk 9]
```diff
@@ -104,6 +104,314 @@ share a bound.
...
+
+Alter the CBMC symbolic execution engine to create a dynamic symbolic execution tool similar
+to [KLEE](https://klee.github.io/). Possible features include:
+
+* Distributed and incremental support using the caching scheme here
+* A feature to run BMC as you go to explore the traces 'around' the execution traces
+* Use path exploration as the witness generation part of an ACDL process.
+* Perform precondition inference, possibly using ACDL.
+
+See the work Björn and Alex have done with path-symex. (?)
```
Comment (tautschnig) [inline MINI-PROJECTS.md]: We do now have a single-path symex mode, but the ideas above might still be relevant.

File: MINI-PROJECTS.md [hunk 10]
```diff
@@ -104,6 +104,314 @@ share a bound.
...
+
+## Interact with GDB
+
+GDB implements a client-server protocol with a GDB instance acting as a server and the
```
Comment (tautschnig) [inline MINI-PROJECTS.md:183]: There is an open PR on that one (#3983).

**Review by tautschnig — COMMENTED** [review]
I think this should get a thorough review from @chrisr-diffblue @peterschrammel and whoever else has a good overview of what the status of various bits of work is.

File: MINI-PROJECTS.md (hunk 1 above)
Comment (peterschrammel) [reply MINI-PROJECTS.md]: No, #1501 is about Minisat options being passed through, #3243 seems similar.
Cesar is definitely not working on that - so, remove that line.

File: MINI-PROJECTS.md [hunk 11]
```diff
@@ -104,6 +104,314 @@ share a bound.
...
+that become trivially true should be eliminated.
+
+### Make SAMATE Tests Work
+
+Read [this paper](https://www.sciencedirect.com/science/article/abs/pii/S0950584913000384) and
+make the tests mentioned in there work.
+
+### Error Explanation
+
+Re-implement the algorithm described in [this paper](http://www.kroening.com/papers/STTT-error-explanation-2005.pdf)
```
Comment (peterschrammel) [inline MINI-PROJECTS.md:147]: This is partly implemented in the fault localizer.

File: MINI-PROJECTS.md (hunk 5 above)
Comment (peterschrammel) [reply MINI-PROJECTS.md]: I don't think so. I suggest to remove this comment.

File: MINI-PROJECTS.md (hunk 7 above)
Comment (peterschrammel) [reply MINI-PROJECTS.md]: Yes... can be removed.

File: MINI-PROJECTS.md (hunk 8 above)
Comment (peterschrammel) [reply MINI-PROJECTS.md:169]: I'd propose a project to implement an interpreter using `goto-symex`.

File: MINI-PROJECTS.md [hunk 12]
```diff
@@ -104,6 +104,314 @@ share a bound.
...
+
+One option would be to convert the trace to a GDB script but this misses some of the value
+as it would be possible to query they GDB server as the program runs to check that values
+really are assigned correctly. This will catch the first point at which the model and the
+real system diverge and allows checked 'resimulation' of the traces on a 'real' platform.
+If the goto-binary contains the ELF binary then this could be used directly.
+
+### Interpreter as a GDB server
+
+An enabling tech, this would allow us to run the interpretter as a debugger back end. Combine
```
Comment (peterschrammel) [inline MINI-PROJECTS.md]: interpreter

File: MINI-PROJECTS.md (hunk 9 above)
Comment (peterschrammel) [reply MINI-PROJECTS.md]: Unclear what the points above mean, but there is certainly work to be done around the single-path symex, including robustness, efficiency and search strategies.

File: MINI-PROJECTS.md [hunk 13]
```diff
@@ -104,6 +104,314 @@ share a bound.
...
+* Use path exploration as the witness generation part of an ACDL process.
+* Perform precondition inference, possibly using ACDL.
+
+See the work Björn and Alex have done with path-symex. (?)
+
+## Interact with GDB
+
+GDB implements a client-server protocol with a GDB instance acting as a server and the
+GUI as a client. (The documentation for that protocol may be found [here](http://www.sourceware.org/gdb/current/onlinedocs/gdb/GDB_002fMI.html).)
+By implementing this protocol in various place within CProver, there are various
```
Comment (peterschrammel) [inline MINI-PROJECTS.md]: places

File: MINI-PROJECTS.md (hunk 10 above)
Comment (peterschrammel) [reply MINI-PROJECTS.md:183]: I don't think that covers all the points below.

File: MINI-PROJECTS.md [hunk 14]
```diff
@@ -104,6 +104,314 @@ share a bound.
...
+of real functions and accounting for this. This would also be necessary for some of the
+preceeding ideas.
+
+### Trace as a GDB server
+
+Implement a GDB server that uses a trace to 'execute' the program. This allows any GUI
+debugger that can talk to GDB to be used to graphically explore the trace. Particularly
+valuable for concurrent programs in which just reading the trace could be very hard.
+
+Pascal has something like this implemented for Eclipse. (?)
```
Comment (peterschrammel) [inline MINI-PROJECTS.md]: An implementation of that is in the trace debugger here: https://github.com/diffblue/eclipse-cbmc

File: MINI-PROJECTS.md [hunk 15]
```diff
@@ -104,6 +104,314 @@ share a bound.
...
+If the goto-binary contains the ELF binary then this could be used directly.
+
+### Interpreter as a GDB server
+
+An enabling tech, this would allow us to run the interpretter as a debugger back end. Combine
+with the trace as a GDB client, this would allow resimulation of traces, which would be useful
+for catching encoding bugs in CProver and ideally should be run after the decode from the solver
+all of the time.
+
+Michael has a partial implementation of this. (?)
```
Comment (peterschrammel) [inline MINI-PROJECTS.md:226]: Really, @tautschnig ?

File: MINI-PROJECTS.md [hunk 16]
```diff
@@ -104,6 +104,314 @@ share a bound.
...
+### CryptoMinisat
+
+Try [CryptoMinisat](https://github.com/msoos/cryptominisat) vs Minisat.
+
+### k-liveness
+
+Implement [k-liveness](http://www.cs.utexas.edu/~hunt/FMCAD/FMCAD12/013.pdf)
+for software-liveness checking.
+
+Hongyi is now looking at this (?)
```
Comment (peterschrammel) [inline MINI-PROJECTS.md]: Not anymore - remove this.

File: MINI-PROJECTS.md [hunk 17]
```diff
@@ -104,6 +104,314 @@ share a bound.
...
+Valgrind and the Linux kernel both have GDB servers - what could we do with these?
+
+## Static Detection of Data Races
+
+It would be good to have a tool that checks for data races in a similar fashion to
+DRD / Helgrind. The current pointer analysis can over approximate shared variables,
+so possibly some combination of this, trace generation, DRD / Helgrind and then
+refinement of which traces to pick? Perhaps this is even something ACDL like?
+
+### CORAL / Q
```
Comment (peterschrammel) [inline MINI-PROJECTS.md:262]: I'd propose to implement something like stratified inlining, for example.

File: MINI-PROJECTS.md [hunk 18]
```diff
@@ -104,6 +104,314 @@ share a bound.
...
+### k-liveness
+
+Implement [k-liveness](http://www.cs.utexas.edu/~hunt/FMCAD/FMCAD12/013.pdf)
+for software-liveness checking.
+
+Hongyi is now looking at this (?)
+
+### Known Bugs on the Web
+
+Turn the list of `KNOWNBUG`S from the CBMC regressions into a webpage of known bugs
```
Comment (peterschrammel) [inline MINI-PROJECTS.md]: Webpage means github issue tracker

Comment (peterschrammel) [reply MINI-PROJECTS.md]: We should also use the information on fixed bugs that we get from CI.

File: MINI-PROJECTS.md [hunk 19]
```diff
@@ -104,6 +104,314 @@ share a bound.
...
+Build instrumenters for various coding standards, e.g. AutoSar, [MISRA-C](https://en.wikipedia.org/wiki/MISRA_C),
+CERT-C, [HIC++](https://www.perforce.com/resources/qac/high-integrity-cpp-coding-standard), etc.
+
+### Front-End Testing
+
+Run the regressions mentioned [here](https://github.com/kframework/c-semantics/tree/master/tests) and
+the [GCC C Torture tests](https://gcc.gnu.org/onlinedocs/gccint/C-Tests.html) with `goto-cc` and a
+`goto-program` interpreter (mild dependency on the interpreter-related task above).
+
+### Unspecified C Behaviours
```
Comment (peterschrammel) [inline MINI-PROJECTS.md:291]: That's a subset of the Coding Standards Checker.

File: MINI-PROJECTS.md [hunk 20]
```diff
@@ -104,6 +104,314 @@ share a bound.
...
+
+The refinement option enables a tactic of under and over approximation to try to solve the
+generated formulae more quickly. There are a large number of possible options – which operations
+to approximate, how much to under or over approximate them, how much to add back if this doesn't
+work, etc. Rather than trying to manually pick the best strategy from these, it would be good
+to machine learn them.
+
+It would probably be worth asking Nando de Freitas' advice on the machine learning side of things.
+
+Ruben is now looking at this one. (?)
```
Comment (peterschrammel) [inline MINI-PROJECTS.md]: Not anymore - remove this line.

File: MINI-PROJECTS.md [hunk 21]
```diff
@@ -104,6 +104,314 @@ share a bound.
...
+Peter may have some thoughts on this. (?)
+
+### Java Memory Model
+
+CBMC's Java front-end is currently in development and will support the full Java bytecode
+instruction set soon. Beyond that, it would be great to support some more advanced aspects
+of the Java programming language semantics, such as the Java Memory Model. This will include
+adding models for a subset of the Java concurrency packages in the Java Runtime Library.
+
+Jade Alglave may have some additional input on this. (?)
```
Comment (peterschrammel) [inline MINI-PROJECTS.md]: Cesar and Kurt have done work on that.

File: MINI-PROJECTS.md [hunk 22]
```diff
@@ -104,6 +104,314 @@ share a bound.
...
+This may imporve the quality of counter-examples (although at the cost of potentially missing
+some bugs, although not if combined with the previous technique).
+
+### C++ Exceptions
+
+C++ has annotations for what exceptions a function can throw. It would be good to be
+able to check / generate / minimise these and it should be possible to do this with
+a reasonably simple abstract domain. The CVC4 project would be interested in using this if it exists.
+
+Peter may have some thoughts on this. (?)
```
Comment (peterschrammel) [inline MINI-PROJECTS.md]: An old prototype of something similar is here: https://github.com/peterschrammel/cbmc/commits/progres-exceptions

File: MINI-PROJECTS.md [hunk 23]
```diff
@@ -104,6 +104,314 @@ share a bound.
...
+
+Also look at Google's Native Client, and talk to Matt, who has tried Valgrind.
+
+Furthermore, talk to Michael, who has got a student who made some progress on this.
+
+And maybe have a read of Automated synthesis of symbolic instruction encodings from I/O samples.
+
+### Branch Prediction
+
+Implement branch prediction algorithms in path-symeex.
```
Comment (peterschrammel) [inline MINI-PROJECTS.md]: path-symex

File: MINI-PROJECTS.md [hunk 24]
```diff
@@ -104,6 +104,314 @@ share a bound.
...
+
+### x86 to Goto
+
+Look at various efforts to model x86 assembler, e.g., Andrew Kennedy,
+“Formalizing .EXEs, .DLLs, and all that”, Gang Tan, “Reusable tools for formal modeling of machine code”,
+and check whether this can turn into a translator from x86 to goto binary.
+
+Also look at Google's Native Client, and talk to Matt, who has tried Valgrind.
+
+Furthermore, talk to Michael, who has got a student who made some progress on this.
```
Comment (peterschrammel) [inline MINI-PROJECTS.md]: @tautschnig, yes?

**Comment by TGWDB** [comment]
@NlightNFotis Is this still going to be worked on and maybe merged?

File: MINI-PROJECTS.md [hunk 25]
```diff
@@ -104,6 +104,314 @@ share a bound.
   those bounds is reached. Good, thoughtful test cases are essential
   here!
 
+### Instrumentation to Check Pointer Alignment
+
+Some architectures only allow load instructions / pointer dereferencing when the
+address is a multiple of the word length (i.e. 4 bytes or 8 bytes). This project
+would add a new instrumentation option for `cbmc` and `goto-instrument` that checks
+each pointer before dereference and makes sure it's correctly aligned.
+
```
Comment (martin-cs) [inline MINI-PROJECTS.md:107]: This is a good way to get into `goto_check` and how instrumentation of asserts work.

File: MINI-PROJECTS.md [hunk 26]
```diff
@@ -104,6 +104,314 @@ share a bound.
...
+  int x;
+  int *ptr;
+  ptr=&x;
+  *ptr=42;
+  if(x < 42)
+    return 1;
+  return 0;
+}
+```
+
```
Comment (martin-cs) [inline MINI-PROJECTS.md:136]: This should be pretty straight-forward to do with `goto-analyser --vsd` and might even be a good intro to that code.

File: MINI-PROJECTS.md (hunk 4 above)
Comment (martin-cs) [reply MINI-PROJECTS.md]: `goto-analyzer --simplify out.gb --vsd ...` should do everything except the concurrency aspect.  Good concurrent abstract interpretation is non-trivial.

File: MINI-PROJECTS.md (hunk 8 above)
Comment (martin-cs) [reply MINI-PROJECTS.md:169]: Using the abstract interpreter would also be possible.

File: MINI-PROJECTS.md (hunk 9 above)
Comment (martin-cs) [reply MINI-PROJECTS.md]: I believe the original suggestion (I have a feeling I wrote a number of these) was using ACDL-like reasoning to "learn" invariants from failed symex runs (using UNSAT cores and the like) and add these in to the process, giving something like IMPACT but without the covering relations.

File: MINI-PROJECTS.md [hunk 27]
```diff
@@ -104,6 +104,314 @@ share a bound.
...
+refinement of which traces to pick? Perhaps this is even something ACDL like?
+
+### CORAL / Q
+
+Build something similar to the Q tool in SLAM. (?)
+
+### CryptoMinisat
+
+Try [CryptoMinisat](https://github.com/msoos/cryptominisat) vs Minisat.
+
```
Comment (martin-cs) [inline MINI-PROJECTS.md]: I believe this is now implemented.

File: MINI-PROJECTS.md [hunk 28]
```diff
@@ -104,6 +104,314 @@ share a bound.
...
+### Unspecified C Behaviours
+
+Build instrumentation to check for unspecified behaviours in C, as per Section J.2
+(and maybe also J.3) of the C standard. Some of them may be possible to handle
+entirely in the front-end (for example, catching `x = x++` as modifying a variable
+twice between sequence points could be done in side effect removal in ansi-c), while
+others may be best to be instrumentation (for example handling the unspecified nature
+of the order of evaluation of function arguments).
+
+Talk to Michael Tautschnig or Martin Brain for this item.
```
Comment (martin-cs) [inline MINI-PROJECTS.md]: I am, of course, happy to chat to anyone about any of these but I don't recall anything specific about this one so maybe I shouldn't be mentioned directly?

File: MINI-PROJECTS.md [hunk 29]
```diff
@@ -104,6 +104,314 @@ share a bound.
...
+
+### Machine Learn Refinement settings
+
+The refinement option enables a tactic of under and over approximation to try to solve the
+generated formulae more quickly. There are a large number of possible options – which operations
+to approximate, how much to under or over approximate them, how much to add back if this doesn't
+work, etc. Rather than trying to manually pick the best strategy from these, it would be good
+to machine learn them.
+
+It would probably be worth asking Nando de Freitas' advice on the machine learning side of things.
```
Comment (martin-cs) [inline MINI-PROJECTS.md]: Although I am sure that he has many insightful and helpful things to say on the subject, the specific link between him and the bulk of the CPROVER developers (being at the same institution, one floor different, etc.) has gone so maybe we shouldn't explicitly name check him here?

File: MINI-PROJECTS.md [hunk 30]
```diff
@@ -104,6 +104,314 @@ share a bound.
...
+### Branch Prediction
+
+Implement branch prediction algorithms in path-symeex.
+
+### An abstract interpreter for C
+
+There's a lack for a robust, open source, abstract interpreter for C/C++. CPROVER contains most of the
+components for it, the key thing that would be needed is a front-end and some testing, plus possibly
+some more domains.
+
```
Comment (martin-cs) [inline MINI-PROJECTS.md:347]: This is `goto-analyzer`.  We can remove this project suggestion -- turns out to not be so mini after all!

File: MINI-PROJECTS.md [hunk 31]
```diff
@@ -104,6 +104,314 @@ share a bound.
...
+### GCC Torture Tests
+
+GCC has a set of 'torture tests' which are intended to stress test and find bugs in their C parser
+and front-end. It would be good to run these through goto-cc / add them to the nightly testing and
+fix any bugs that arise. (Possible overlap with “Front-End Testing” above.). The same is probably
+also true for CSmith and other compiler fuzzing tools.
```
Comment (martin-cs) [inline MINI-PROJECTS.md]: @tautschnig has implemented the CSmith part.

**Review by martin-cs — APPROVED** [review]
It would be great to see this patch merged but it does need a few updates.

**Comment by NlightNFotis** [comment]
@TGWDB Yes, I intend to chase it up and clean this up soon (as soon as I get some time).

@martin-cs Thanks for the suggestions and the comments, will incorporate them when I find some time 👍🏻

**Comment by NlightNFotis** [comment]
Hi @peterschrammel and @tautschnig, could you take a look at this at your latest convenience?

I've implemented most of the comments, and changed the file a bit to conform to markdown listing standards.

Other than this, it looks good to go to me, what do you think?

**Review by peterschrammel — APPROVED** [review]

**Pull request merged**